### DISCUSSION TITLE
how to use this onModelBeforeUpdate in js

### DISCUSSION THREAD

**Discussion by apujayed** [pocketbase/pocketbase#3443] [Q&A] [answered]
i want to do a task before update the data how can i do that by using onModelBeforeUpdate.make an example.thanks

**Comment by ganigeorgiev** [comment]
I'm not sure I follow. The question is too vague and I don't understand what the example should illustrate.

If you want to modify a record field before persisting it in the db, this could look something like:

```js
onModelBeforeUpdate((e) => {
    e.model.set("someField", 123)
}, "yourCollection")
```

`e.model` in the above is a Record instance. If you haven't already, you can also check the [Record operations](https://pocketbase.io/docs/js-records/) guides.

If that doesn't help, please elaborate on your use case (_eg. describe your db structure and what needs to be done_) and/or provide a pseudo-code with the desired behavior.

**Reply by apujayed** [reply]
my requirements is whenever user request to update a field in my collection than it check before the update every request that must be greater value than the collection existing field value.in frontend i use react js..now how can i do this task?

**Reply by ganigeorgiev** [reply]
You must add the above in file ending with `.pb.js` inside `pb_hooks` directory (eg. `pb_hooks/main.pb.js`) and restart the executable.

For more details, make sure to check the [Extend with JS - Overview](https://pocketbase.io/docs/js-overview/).

**Reply by ganigeorgiev** [reply]
_Note to get the old/original record field value you can call `e.model.originalCopy().get("yourField")` (_see [Record operations](https://pocketbase.io/docs/js-records/#getset-record-fields) for overview of the available methods_)._

**Reply by apujayed** [reply]
i dont get that pb_hook directory.where i need to keep this?can you give me a clear scenario what i need to do..i mean how can i use this hook?i have to make another server?using node js or anything?

**Reply by ganigeorgiev** [reply]
You don't need another server. The `pb_hooks` directory need to be created next to the PocketBase executable (similar to the `pb_migrations` and `pb_data`).

Again, please make sure to read [Extend with JS - Overview](https://pocketbase.io/docs/js-overview/) as there are some caveats related to the embedded JS engine.

**Reply by apujayed** [reply]
okay thanks let me try

**Reply by apujayed** [reply]
how can i get the value which will be updated?i mean before update i need to check if e.model.originalCopy().get("yourField")<user input if match than update otherwise not how to implement this condition

**Reply by ganigeorgiev** [reply]
`e.model.get("yourField")` will return the user submitted value that it is not yet persisted.

_Side-note: if all you want is to compare the submitted and original value of a single field then you can do that also with the Collection Update API rule: `@request.data.yourField > yourField`_.

**Comment by apujayed** [comment]
`onModelBeforeUpdate((e) => {
if(e.model.originalCopy().get("bag")<e.model.get("bag"))
   {
    console.log('true');
   }else{
    console.log('false');
   }
})`
 here if true only than update else throw an error how to implement this?

**Reply by ganigeorgiev** [reply, marked as answer]
You can throw a standard JS error, eg. `throw new Error("something went wrong...")`.

If you want to return to the client a formatted API error, you can attach to one of the `*Request` hooks and return an `ApiError` as shown in https://pocketbase.io/docs/js-routing/#error-response. For example:

```js
// pb_hooks/main.pb.js

onRecordBeforeUpdateRequest((e) => {
    const oldBag = e.record.originalCopy().get("bag");
    const newBag = e.record.get("bag");

    if (newBag > oldBag) {
        throw new BadRequestError("Something went wrong...", {
            "bag":  new ValidationError("invalid_bag_value", "Bag must be less than " + oldBag),
        })
    }
}, "yourCollection");
```
_(don't forget to replace `yourCollection` with your actual collection name)_

**Reply by apujayed** [reply]
Thank you.its work for me🙂
